compute_tools/src/vm_monitor/cgroup.rs

```diff
@@ -0,0 +1,777 @@
+            MemoryEvent::High => f.write_str("high"),
+            MemoryEvent::Max => f.write_str("max"),
+            MemoryEvent::Oom => f.write_str("oom"),
+            MemoryEvent::OomKill => f.write_str("oom_kill"),
+            MemoryEvent::OomGroupKill => f.write_str("oom_group_kill"),
+        }
+    }
+}
+
+/// An event that the `CgroupWatcher` responds to.
+///
+/// The `CroupWatcher`'s job primarily consists of reacting to these events.
+/// See the `CgroupWatcher` docs and associated impl items for more information.
+#[derive(Debug, Clone)]
+enum EventKind {
+    /// We were upscaling to the contained allocation
+    Upscale(Resources),
+
+    /// Number of memory.highs
+    MemoryHigh(u64),
+}
+
+/// An extension to a stream that makes it peekable.
+///
+/// Note: this should probably be replaced using futures::StreamExt::Peekable
+struct PeekableStream<T> {
+    stream: BoxStream<'static, T>,
+    peek: Option<T>,
+}
+
+impl<T> Debug for PeekableStream<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PeekableStream")
+            .field("stream", &"{ .. }")
+            .field("peek", &self.peek)
+            .finish()
+    }
+}
+
+impl<T> PeekableStream<T> {
+    pub fn new<S>(stream: S) -> Self
+    where
+        S: Stream<Item = T> + Send + 'static,
+    {
+        Self {
+            stream: futures::StreamExt::boxed(stream),
+            peek: None,
+        }
+    }
+
+    /// Peek the stream, but don't await a new element if there isn't currently
+    /// one in that we've already peeked.
+    pub fn peek_eager(&mut self) -> Option<&T> {
+        // Eagerly check for next item if the peek isn't initalized
+        if self.peek.is_none() {
+            self.peek = self
+                .stream
+                .next()
+                .now_or_never()
+                .expect("failed to read from stream");
+        }
+        self.peek.as_ref()
+    }
+
+    /// Peek the stream.
+    // NOTE: currently unused, but potentially useful for the future. Also good
+    // to distinguish between `peek` and `peek_eager`
+    #[allow(unused)]
+    pub async fn peek(&mut self) -> Option<&T> {
+        if self.peek.is_none() {
+            self.peek = self.stream.next().await;
+        }
+        return self.peek.as_ref();
+    }
+}
+
+impl<T> Stream for PeekableStream<T>
+where
+    T: Unpin,
+{
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
+        // first check if there's something we've already pulled from the stream
+        // in the peek
+        if let Some(data) = self.as_mut().peek.take() {
+            return Poll::Ready(Some(data));
+        }
+        // otherwise we actually need to pull something from the stream
+        let next = self.stream.next();
+        tokio::pin!(next);
+        Future::poll(Pin::new(&mut next), cx)
```

Comment:
This looks odd; this is essentially a manual implementation of https://docs.rs/futures/latest/futures/future/fn.poll_immediate.html

But yes, much better to use futures::StreamExt::Peekable if streams are even needed.